### PR TITLE
Implement simulator

### PR DESCRIPTION
**Pull request by joaobrittoneto**
It would simulate the usbl and the acoustic link.
The Task will simulate the driver and the interface with
the auv_uwmodem_data and auv_uwmodem_commands

### CODE REVIEW

**Comment by joaobrittoneto** [comment]
@doudou,
I intend to simulate the acoustic link (UWPhysicalMean) and the driver (Task).
One doubt about the delay of transmission. How could it be implemented in such a way that simulation could run with log files and accelerated speed, i.e. don't rely on the present time and it feed rate?

**Comment by joaobrittoneto** [comment]
@doudou,
Any comment here? May I merge?

**Comment by doudou** [comment]
Sorry ... been neglecting my github duties ...

File: uwmodem_simulation.orogen [hunk 1]
```diff
@@ -4,16 +4,23 @@ import_types_from "uwmodem_simulationTypes.hpp"
...
     # This will affect the rate of transmission data in raw_data_output
     # For SERIAL it is 2 bytes/Rawpacket. For ETHERNET it is a direct transmission.
-    property "interface", "/usbl_evologics/InterfaceType"
+    property "receiver_interface", "/usbl_evologics/InterfaceType", :ETHERNET
     # Set average distance in meters between modems.
     # It affects the transfer delay. It does not affect the rate of lost data.
     property "distance", "double", 300
+    # Set probability of good transmission in %
+    # It will affect the lost of data
+    property "probability", "int", 90
```
Comment (doudou) [inline uwmodem_simulation.orogen]: Percentage is a bit like degrees .... Don't use them. Use probabilities in [0, 1](floating point)

**Comment by doudou** [comment]
Looks really good. Just the little comment on probability representation in the properties.

What do you have in mind w.r.t. transmission delay and tests ? What specific test do you have in mind ?

**Comment by joaobrittoneto** [comment]
Thanks.
I use a lot of base::Time::now for counting the travel time of packets/IM between modems and bitrate control.
The problem is that I can not run the tests fast and for the case of several hundreds/thousands packets it may takes ages (for checking the good transmission rate for example). 
Also I wonder how it could behave if we feed the simulation with a log file and old timestamp. If the  inconsistency of timestamp in input and output data could be a problem.

**Comment by doudou** [comment]
> I use a lot of base::Time::now for counting the travel time of packets/IM between modems and bitrate control. The problem is that I can not run the tests fast and for the case of several hundreds/thousands packets it may takes ages (for checking the good transmission rate for example). 

Can't you make the rate limit much higher (for the purpose of the tests) and, which would comparatively allow you to make the test shorter ?

> Also I wonder how it could behave if we feed the simulation with a log file and old timestamp. If the inconsistency of timestamp in input and output data could be a problem.

This is a general, and generally not fixed, problem in Rock. I've sketched the infrastructure work needed to allow for having logical times (i.e. in-simulation), but never had the time to implement it.

In your particular case, it will depend whether you use any time information that is within the data itself, and/or if you timestamp the output _and_ we use this timestamp further down the chain.

**Comment by joaobrittoneto** [comment]
@doudou,
I made the output's timestamp be related with input's timestamp. 
The timestamp is not transmitted direct, but I think it would make sense making the input/output's timestamp correlated, even if I'm not sure how, or by who, the component will be used.

**Comment by doudou** [comment]
Hey. Looks good overall. We'll figure out the rest when we start using it more extensively I guess

**Comment by joaobrittoneto** [comment]
Merging

**Pull request merged**